internal/controllers/phase_ensure_wanted_namespaces.go

```diff
@@ -0,0 +1,162 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	k8sApiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+// Ensure existence of namespaces specified in the given addon resource
+// returns a bool that signals the caller to stop reconilation and retry later
+func (r *AddonReconciler) ensureWantedNamespaces(
+	ctx context.Context, addon *addonsv1alpha1.Addon) (stopAndRetry bool, err error) {
+	var ensuredNamespaces []*corev1.Namespace
+	var collidedNamespaces []string
+
+	for _, namespace := range addon.Spec.Namespaces {
+		ensuredNamespace, err := r.ensureNamespace(ctx, addon, namespace.Name)
+		if err != nil {
+			if errors.Is(err, errNotOwnedByUs) {
+				collidedNamespaces = append(collidedNamespaces, namespace.Name)
+				continue
+			}
+
+			return false, err
+		}
+
+		ensuredNamespaces = append(ensuredNamespaces, ensuredNamespace)
+	}
+
+	if len(collidedNamespaces) > 0 {
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:   addonsv1alpha1.Available,
+			Status: metav1.ConditionFalse,
+			Reason: "CollidedNamespaces",
+			Message: fmt.Sprintf(
+				"Namespaces with collisions: %s",
+				strings.Join(collidedNamespaces, ", ")),
+			ObservedGeneration: addon.Generation,
+		})
+		addon.Status.ObservedGeneration = addon.Generation
+		addon.Status.Phase = addonsv1alpha1.Pending
+		err := r.Status().Update(ctx, addon)
+		if err != nil {
+			return false, err
+		}
+		// collisions occured: signal caller to stop and retry
+		return true, nil
+	}
+
+	var unreadyNamespaces []string
+
+	for _, ensuredNamespace := range ensuredNamespaces {
+		if ensuredNamespace.Status.Phase != corev1.NamespaceActive {
+			unreadyNamespaces = append(unreadyNamespaces, ensuredNamespace.Name)
+		}
+	}
```

Comment:
Can be inlined into loop in line #26-28 and would result in less indirection by getting rid of `ensuredNamespaces` and does not require to loop over the same list twice.

Also don't put pointers into slices if there is not a very good reason to do so:
https://medium.com/@philpearl/bad-go-slices-of-pointers-ed3c06b8bb41

Reply:
:+1: and thanks for the link!